### DISCUSSION TITLE
Object Water Interaction Discussion

### DISCUSSION THREAD

**Discussion by Revmatek** [crest-ocean/crest#1075] [General]
A few things up front for this, this is more of an open discussion, rather than a request if not only for my own understanding. 

I have experimented with both the ObjectInteraction and SphereInteraction scripts / shaders. I have found and implemented use cases for both.

ObjectInteraction use cases / pros / cons:
Pros:
Dynamic meshes / Unpredictable meshes
Custom shaders / scripts that require mesh data
ObjectInteraction shader variations per use case (same shader changed slightly to accomplish a different goal, used as a template)
High speed custom wake generation (200 kts +)
Cons:
Higher performance impact? (I have not noticed much)
Does not obtain desired realism for boat wakes (specific use case)

SphereInteraction:
Pros:
Dropping a stone into the water
Static / Slow moving boat wakes (< 30 kts)
Known object shapes that can have spheres placed ahead of time at key positions to approximate shape
Cons:
Seems to break down at higher speeds
Noticeable spherical artifacts from motion

For both:
This is maybe a bug, but it seems if you have larger interactions they tend to start a sort of feedback loop and push water into the air, I debugged this at one point and there is a value somewhere deep in the RenderTexture / Shader code that flips and comes up negative causing deformation to generate upward at all times rather than downward. You can see this in "BoatWakes" if you pick one of the SphereIteractions and max out the radius. Is there some reason the size is limited? ObjectInteraction suffers from the same pitfall, but there are ways to work around it will custom code.

I am not ready to stop using ObjectInteraction, but I am also using SphereInteraction for various use cases and they seem like 2 different tools / examples / templates that use the same subset features I am curious if there is perhaps a better way to go about using SphereInteraction.

**Comment by magenta404** [comment]
> This is maybe a bug, but it seems if you have larger interactions they tend to start a sort of feedback loop and push water into the air, I debugged this at one point and there is a value somewhere deep in the RenderTexture / Shader code that flips and comes up negative causing deformation to generate upward at all times rather than downward. You can see this in "BoatWakes" if you pick one of the SphereIteractions and max out the radius. Is there some reason the size is limited? ObjectInteraction suffers from the same pitfall, but there are ways to work around it will custom code.

I tried this and it does create a large bump in the water. Is this was you're referring to? I believe this is normal as the interactions add forces and then all forces are dampened in a separate call. So this could just be case of more forces being added than the simulation can dampen.

Negative values are valid.

**Reply by Revmatek** [reply]
The adjustable strength sure makes a difference in the bump, yes. Really the behavior I am referring to that I was trying to confirm also happens with Spheres is that when the mesh you are uses gets larger, even though the strength is "normal" it will apply positive deformation  for an undetermined amount of time, then for whatever reason it will invert, go negative and start creating upward deformation and never revert back to the "normal" identical positive deformation. It was behavior I found problematic trying to create dynamic waves which did not have a predefined shape.

**Reply by magenta404** [reply]
That might be instability. Do you have a screenshot?

**Reply by Revmatek** [reply]
I do not have a screenshot. I may be able to get some steps to reproduce in the future, but I am busy with some other matters atm.  From memory you can reproduce similar behavior in the boat waves example by maxing out the strength and or making the spheres very large. Another aspect I stumbled upon recently was that the rotation of the sphere affects the deformation inversion in some way... need to look more at it, but keeping it level avoids the issue in some cases (when not very large) when attached to a rigidbody. It's possible that this has something to do with common issues surrounding Euler angles.

**Comment by magenta404** [comment]
> I am not ready to stop using ObjectInteraction, but I am also using SphereInteraction for various use cases and they seem like 2 different tools / examples / templates that use the same subset features I am curious if there is perhaps a better way to go about using SphereInteraction.

The plan is for SWI to replace OWI. The idea is that most shapes can be made by using multiple SWI components. There is a plan to add more procedural shapes like a capsule.

> Seems to break down at higher speeds
> Noticeable spherical artifacts from motion

Are you using the latest version (4.15.2)? Improvements to higher speeds have been made since its introduction.

**Reply by Revmatek** [reply]
> The plan is for SWI to replace OWI. The idea is that most shapes can be made by using multiple SWI components. There is a plan to add more procedural shapes like a capsule.

As long as there is a way to do things procedurally I think that covers the bases. I would just note that there are use cases for exact mesh / vertices that are touching the water to achieve things with modified shaders and scripts. A one off example might be having precise foam in the shape of the points contacting the water, maybe someone wants to smash a custom shape into the water for a cut scene or something, you could approximate that procedurally, but I think that might be less fidelity and /or require processing time for mesh data vs having a shader which applies things using the mesh. If the OWI shader is removed it would mean having to jerry rig something similar to obtain precise effects, even if it is just good as a template for modifying to make other cool mesh effects. I still tend to think removing that would make it more difficult for others to evaluate whatever mechanics they are trying to produce, not necessarily boat wakes or even realize that could be done.

> Are you using the latest version (4.15.2)? Improvements to higher speeds have been made since its introduction.

I'm on an older version that I should probably upgrade. I did not know about the improvements for higher speeds, I will check that out, thanks! I did check this in the boat wakes scenes in the latest version just now and set the translation speed to 100 (m/s?) its seems about the speed Id expect and you definitely still get circles, you can play around with the settings a bit and it changes but never looks like it does moving at much slower rates (9), it seems to start breaking down around 17.57 for the smallest sphere (0.5m) and there are less and less circle artifacts visible as sizes increase between the examples in the scenes. It does seems very similar to what I was experiencing.

I was able to get more precision with a modified Interaction script that uses Rigidbody velocity in FixedUpdate rather than LateUpdate instead of a point to point calculation for object interactions, which it now occurs to me I may be able to do something similar for sphere interaction. I will experiment with that.

Is there a version file in the project?

**Reply by magenta404** [reply]
> A one off example might be having precise foam in the shape of the points contacting the water, maybe someone wants to smash a custom shape into the water for a cut scene or something

We have foam inputs for that use case which do support custom meshes albeit it has some difficulties with setup.

> Is there a version file in the project?

Version should be in _Constants.cs_.

**Reply by Revmatek** [reply]
Ok thanks, I think I got some good starting points for what I was trying to do for SphereInteractions. I will keep an eye out for procedural updates in the future.
